File: src/android/app/src/main/java/com/memoryreel/constants/ErrorConstants.java
package com.memoryreel.constants;

import androidx.annotation.IntDef;
import androidx.annotation.StringDef;
import androidx.annotation.Keep;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

/**
 * Standardized error constants, types and messages for the MemoryReel Android application.
 * Provides type-safe error handling with Android-specific annotations.
 */
@Keep
public final class ErrorConstants {

    /**
     * HTTP status codes with type-safe integer definitions
     */
    public static final class HTTP_STATUS {
        public static final int OK = 200;
        public static final int CREATED = 201;
        public static final int BAD_REQUEST = 400;
        public static final int UNAUTHORIZED = 401;
        public static final int FORBIDDEN = 403;
        public static final int NOT_FOUND = 404;
        public static final int CONFLICT = 409;
        public static final int TOO_MANY_REQUESTS = 429;
        public static final int INTERNAL_SERVER_ERROR = 500;
        public static final int SERVICE_UNAVAILABLE = 503;

        private HTTP_STATUS() {} // Prevent instantiation
    }

    /**
     * Error type constants for categorizing different types of errors
     */
    public static final class ERROR_TYPES {
        public static final String VALIDATION_ERROR = "VALIDATION_ERROR";
        public static final String AUTHENTICATION_ERROR = "AUTHENTICATION_ERROR";
        public static final String AUTHORIZATION_ERROR = "AUTHORIZATION_ERROR";
        public static final String RATE_LIMIT_ERROR = "RATE_LIMIT_ERROR";
        public static final String DATABASE_ERROR = "DATABASE_ERROR";
        public static final String AI_SERVICE_ERROR = "AI_SERVICE_ERROR";
        public static final String STORAGE_ERROR = "STORAGE_ERROR";
        public static final String SERVER_ERROR = "SERVER_ERROR";
        public static final String NETWORK_ERROR = "NETWORK_ERROR";
        public static final String MEDIA_ERROR = "MEDIA_ERROR";

        private ERROR_TYPES() {} // Prevent instantiation
    }

    /**
     * Standardized error messages grouped by error categories
     */
    public static final class ERROR_MESSAGES {
        
        public static final class VALIDATION {
            public static final String INVALID_INPUT = "Invalid input data provided";
            public static final String MISSING_FIELD = "Required field is missing";
            public static final String INVALID_FORMAT = "Invalid data format";

            private VALIDATION() {} // Prevent instantiation
        }

        public static final class AUTH {
            public static final String INVALID_CREDENTIALS = "Invalid email or password";
            public static final String TOKEN_EXPIRED = "Authentication token has expired";
            public static final String INSUFFICIENT_PERMISSIONS = "Insufficient permissions for this action";

            private AUTH() {} // Prevent instantiation
        }

        public static final class RATE_LIMIT {
            public static final String TOO_MANY_REQUESTS = "Too many requests, please try again later";
            public static final String API_LIMIT_EXCEEDED = "API rate limit exceeded";

            private RATE_LIMIT() {} // Prevent instantiation
        }

        public static final class MEDIA {
            public static final String UPLOAD_FAILED = "Failed to upload media";
            public static final String DOWNLOAD_FAILED = "Failed to download media";
            public static final String INVALID_FORMAT = "Unsupported media format";
            public static final String PROCESSING_FAILED = "Failed to process media";

            private MEDIA() {} // Prevent instantiation
        }

        public static final class NETWORK {
            public static final String CONNECTION_ERROR = "No internet connection";
            public static final String TIMEOUT = "Request timed out";
            public static final String SERVER_UNREACHABLE = "Server is unreachable";

            private NETWORK() {} // Prevent instantiation
        }

        public static final class SERVER {
            public static final String INTERNAL_ERROR = "Internal server error occurred";
            public static final String SERVICE_UNAVAILABLE = "Service temporarily unavailable";
            public static final String DATABASE_ERROR = "Database operation failed";

            private SERVER() {} // Prevent instantiation
        }

        private ERROR_MESSAGES() {} // Prevent instantiation
    }

    /**
     * Type-safe annotation for HTTP status codes
     */
    @Keep
    @IntDef({
        HTTP_STATUS.OK,
        HTTP_STATUS.CREATED,
        HTTP_STATUS.BAD_REQUEST,
        HTTP_STATUS.UNAUTHORIZED,
        HTTP_STATUS.FORBIDDEN,
        HTTP_STATUS.NOT_FOUND,
        HTTP_STATUS.CONFLICT,
        HTTP_STATUS.TOO_MANY_REQUESTS,
        HTTP_STATUS.INTERNAL_SERVER_ERROR,
        HTTP_STATUS.SERVICE_UNAVAILABLE
    })
    @Retention(RetentionPolicy.SOURCE)
    public @interface HttpStatus {}

    /**
     * Type-safe annotation for error types
     */
    @Keep
    @StringDef({
        ERROR_TYPES.VALIDATION_ERROR,
        ERROR_TYPES.AUTHENTICATION_ERROR,
        ERROR_TYPES.AUTHORIZATION_ERROR,
        ERROR_TYPES.RATE_LIMIT_ERROR,
        ERROR_TYPES.DATABASE_ERROR,
        ERROR_TYPES.AI_SERVICE_ERROR,
        ERROR_TYPES.STORAGE_ERROR,
        ERROR_TYPES.SERVER_ERROR,
        ERROR_TYPES.NETWORK_ERROR,
        ERROR_TYPES.MEDIA_ERROR
    })
    @Retention(RetentionPolicy.SOURCE)
    public @interface ErrorType {}

    private ErrorConstants() {} // Prevent instantiation
}

File: src/ios/MemoryReel/Constants/ErrorConstants.swift
//
// ErrorConstants.swift
// MemoryReel
//
// Defines standardized error constants, error types, and error messages for consistent error handling
// across the MemoryReel iOS application.
//
// Foundation version: iOS 14.0+
//

import Foundation

/// A utility class providing centralized access to error-related constants for the MemoryReel application.
public final class ErrorConstants {
    
    // MARK: - Error Types
    
    /// Standardized error type identifiers for error classification
    public struct ErrorType {
        public static let validation = "VALIDATION_ERROR"
        public static let authentication = "AUTHENTICATION_ERROR"
        public static let authorization = "AUTHORIZATION_ERROR"
        public static let rateLimit = "RATE_LIMIT_ERROR"
        public static let network = "NETWORK_ERROR"
        public static let storage = "STORAGE_ERROR"
        public static let aiService = "AI_SERVICE_ERROR"
        public static let mediaProcessing = "MEDIA_PROCESSING_ERROR"
        public static let server = "SERVER_ERROR"
        public static let biometric = "BIOMETRIC_ERROR"
    }
    
    // MARK: - HTTP Status Codes
    
    /// Standard HTTP status codes for API responses
    public struct HTTPStatus {
        // Success codes (2xx)
        public static let ok = 200
        public static let created = 201
        public static let accepted = 202
        public static let noContent = 204
        
        // Client error codes (4xx)
        public static let badRequest = 400
        public static let unauthorized = 401
        public static let forbidden = 403
        public static let notFound = 404
        public static let methodNotAllowed = 405
        public static let conflict = 409
        public static let unsupportedMediaType = 415
        public static let tooManyRequests = 429
        
        // Server error codes (5xx)
        public static let internalServerError = 500
        public static let serviceUnavailable = 503
        public static let gatewayTimeout = 504
    }
    
    // MARK: - Error Messages
    
    /// Structured error messages organized by category
    public struct ErrorMessage {
        
        /// Validation-related error messages
        public struct Validation {
            public static let invalidInput = "Invalid input provided"
            public static let requiredField = "Required field missing"
            public static let invalidFormat = "Invalid format"
            public static let invalidFileType = "Unsupported file type"
            public static let fileSizeTooLarge = "File size exceeds maximum limit"
        }
        
        /// Authentication and authorization error messages
        public struct Auth {
            public static let invalidCredentials = "Invalid credentials"
            public static let sessionExpired = "Session has expired"
            public static let unauthorized = "Unauthorized access"
            public static let accountLocked = "Account has been locked"
            public static let invalidToken = "Invalid or expired token"
        }
        
        /// Network-related error messages
        public struct Network {
            public static let noConnection = "No internet connection"
            public static let timeout = "Request timed out"
            public static let serverUnreachable = "Server is unreachable"
            public static let invalidResponse = "Invalid response from server"
            public static let sslError = "Secure connection failed"
        }
        
        /// Storage and media-related error messages
        public struct Storage {
            public static let insufficientSpace = "Insufficient storage space"
            public static let uploadFailed = "Failed to upload media"
            public static let downloadFailed = "Failed to download media"
            public static let deleteFailed = "Failed to delete media"
            public static let corruptedFile = "File is corrupted"
        }
        
        /// AI service-related error messages
        public struct AI {
            public static let processingFailed = "AI processing failed"
            public static let serviceUnavailable = "AI service is temporarily unavailable"
            public static let invalidResponse = "Invalid response from AI service"
            public static let quotaExceeded = "AI service quota exceeded"
            public static let unsupportedContent = "Content type not supported by AI service"
        }
        
        /// Rate limiting error messages
        public struct RateLimit {
            public static let quotaExceeded = "Request quota exceeded"
            public static let tooManyRequests = "Too many requests"
            public static let retryAfter = "Please try again later"
        }
    }
    
    // MARK: - Error Domains
    
    /// Error domains for NSError creation and source identification
    public struct ErrorDomain {
        public static let network = "com.memoryreel.error.network"
        public static let auth = "com.memoryreel.error.auth"
        public static let storage = "com.memoryreel.error.storage"
        public static let validation = "com.memoryreel.error.validation"
        public static let ai = "com.memoryreel.error.ai"
        public static let media = "com.memoryreel.error.media"
        public static let security = "com.memoryreel.error.security"
    }
    
    // MARK: - Initialization
    
    /// Private initializer to prevent instantiation
    private init() {}
}